src/gui/socketapi.cpp

```diff
@@ -608,52 +608,51 @@ void SocketApi::copyUrlToClipboard(const QString &link)
 void SocketApi::command_DOWNLOAD_VIRTUAL_FILE(const QString &filesArg, SocketListener *)
 {
     QStringList files = filesArg.split(QLatin1Char('\x1e')); // Record Separator
-    auto suffix = QStringLiteral(APPLICATION_DOTVIRTUALFILE_SUFFIX);
 
     for (const auto &file : files) {
-        if (!file.endsWith(suffix) && !QFileInfo(file).isDir())
+        auto data = FileData::get(file);
+        auto record = data.journalRecord();
+        if (!record.isValid())
             continue;
-        QString relativePath;
-        auto folder = FolderMan::instance()->folderForPath(file, &relativePath);
-        if (folder) {
-            folder->downloadVirtualFile(relativePath);
-        }
+        if (record._type != ItemTypeVirtualFile && !QFileInfo(file).isDir())
+            continue;
+        if (data.folder)
+            data.folder->downloadVirtualFile(data.folderRelativePath);
     }
 }
 
-/* Go over all the files ans replace them by a virtual file */
+/* Go over all the files and replace them by a virtual file */
 void SocketApi::command_REPLACE_VIRTUAL_FILE(const QString &filesArg, SocketListener *)
 {
     QStringList files = filesArg.split(QLatin1Char('\x1e')); // Record Separator
-    auto suffix = QStringLiteral(APPLICATION_DOTVIRTUALFILE_SUFFIX);
 
+    QSet<Folder *> toSync;
     for (const auto &file : files) {
-        QString relativePath;
-        auto folder = FolderMan::instance()->folderForPath(file, &relativePath);
-        if (!folder)
-            continue;
-        if (file.endsWith(suffix))
+        auto data = FileData::get(file);
+        if (!data.folder)
             continue;
+        auto journal = data.folder->journalDb();
+        auto markForDehydration = [&](SyncJournalFileRecord rec) {
+            if (rec._type != ItemTypeFile)
+                return;
+            rec._type = ItemTypeVirtualFileDehydration;
+            journal->setFileRecord(rec);
+            toSync.insert(data.folder);
+        };
+
```

Comment:
We also need to add the path in the list of path we need to visit.  Otherwise the discovery may skip them.   I think.

Reply:
Yes!

Thinking about it, I assumed that "_invalid_" etags and `avoidReadFromDbOnNextSync` would also encompass local discovery, but at least for the new sync algorithm I don't think it does. (not sure whether it's true for the old code) What do you think about that?

Reply:
avoidReadFromDbOnNextSync would work.  But we also don't need to do server discovery (slow).
folder->_localDiscoveryTracker->addTouchedPath should also work.


Ideally, in the future, we can actually ask the sync engine to just do these files immediatly.